A little generic poptrie implementation

eugeneia opened the pull request:
This branch contains the beginning of an implementation of [Poptrie](http://conferences.sigcomm.org/sigcomm/2015/pdf/papers/p57.pdf). So far it can...

 - build the poptrie lookup structure for prefixes of up to 64 bits
 - look up keys in the poptrie fairly quick using a DynASM based lookup routine
 - be instantiated using a pre-built poptrie structure

Performance on an Intel(R) Xeon(R) CPU E3-1246 v3 @ 3.50GHz (Haswell, Turbo off)

```
PMU analysis (numentries=10000, keysize=32)
build: 0.1290 seconds
lookup: 13217.09 cycles/lookup 28014.35 instructions/lookup
lookup64: 122.94 cycles/lookup 133.22 instructions/lookup
build(direct_pointing): 0.1056 seconds
lookup(direct_pointing): 5519.01 cycles/lookup 11412.01 instructions/lookup
lookup64(direct_pointing): 89.82 cycles/lookup 70.72 instructions/lookup
```

Not implemented yet:

- ~~prefixes/keys of up to 128 bits (i.e., supporting IPv6)~~

Cc @petebristow

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+      | mov offset, 18
+   else
+      -- index, node, offset = 0, nodes[index], 0
+      | mov index, 0
```
Unless you're explicitly preserving flags (which I assume you're not), `xor R, R` is typically better than `mov R, 0` (e.g. because it is a byte or two shorter to encode).

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+      -- v = rshift(key, offset)
+      | mov v, key
+      | mov rcx, offset
+      | shr v, cl
```
If you can assume BMI2 (i.e. Haswell or later IIRC), this `mov`/`mov`/`shr` can probably be expressed as a single `shrx`.

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   -- rax = band(vec, rax)
+   | and rax, vec
+   -- is bit v set in vec?
+   | cmp rax, 0
```
This `cmp` is entirely redundant - the immediately preceding `and` will already have set `ZF=1` if the result was zero.

corsix replied:
Alternatively, the entire sequence `mov`/`mov`/`shl`/`and`/`cmp`/`je` might be better as `bt vec, v`/`jnc`.

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   -- is bit v set in vec?
+   | cmp rax, 0
+   | je >3 -- reached leaf, exit loop
+   -- rax = lshift(2ULL, v) - 1
```
If you can assume BMI2, then this construction of `rax` and subsequent `and` can be replaced by something like:
```
lea rcx, [v+1]
bzhi rax, vec, rcx
```

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   -- return leaves[base + bc - 1]
+   | mov index, dword [node+16] -- nodes[index].base0
+   | add index, eax
+   | sub index, 1
```
This `sub` can be fused into the subsequent memory operand (i.e. `word [leaves+index*2-2]`).

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   | add index, eax
+   | sub index, 1
+   -- assumption: at most ax is set at this point because popcnt(qword) < byte
+   | mov ax, word [leaves+index*2] -- leaves[index]
```
This would probably be better as `movzx eax, ...`. The assumption in the comment is true, but the CPU is not clever enough to know this, and so if you do a `mov` into `ax`, then when `eax` is next used, it'll insert a micro-up to combine `ax` and the previous high 16-bits of `eax` (whereas `movzx` here is the same price as `mov`, and doesn't have the extra micro-op penalty down the line).

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   | popcnt rax, rax
+   -- index = base + bc - 1
+   | mov index, dword [node+20] -- nodes[index].base1
+   | add index, eax
```
There might be value in transposing this `add` and this `sub` - the `add` cannot execute until both the `mov` and the `popcnt` are done, and the `sub` cannot execute until the `add` is done - whereas if transposed, the `sub` could execute in parallel with the `popcnt`.

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+   | lea node, [nodes+index]
+   -- offset = offset + k
+   | add offset, 6
+   | jmp <2 -- loop
```
You could consider rotating the loop to get rid of this unconditional jump on every iteration, i.e. turning:
```
head:
body part 1
jump to tail if condition
body part 2
jmp head
tail:
```
into:
```
jmp middle
head:
body part 2
middle:
body part 1
jump to head if not condition
```
or:
```
body part 1
jump to tail if condition
head:
body part 2
body part 1
jump to head if not condition
tail:
```

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+      -- index = dmap[v]
+      | mov index, dword [dmap+v*4]
+      -- eax = band(index, leaf_tag)
+      | mov eax, 0x80000000
```
Or perhaps `mov eax, index` / `btr eax, 31` / `jnc`, with the bonus of then eliminating the subsequent `mov`/`and`.

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,158 @@
+      local direct_mask = bit.lshift(1ULL, Poptrie.s) - 1
+      -- v = band(key, direct_mask)
+      | mov v, direct_mask
+      | and v, key
```
I'd probably write this `mov`/`and` as `mov v, key` / `and v, direct_mask`, as that can allow for a more compact encoding (`mov r32, imm32` always encodes the immediate as 32 bits, whereas `and r32, imm32` _can_ encode the immediate as 8 bits if it is small enough).

corsix replied:
The other rationale here that I forgot last night: on contemporary architectures, `mov reg, reg` can be zero latency and not require any execution unit, whereas `mov reg, imm` does have latency and does require an execution unit (`op reg, reg` and `op reg, imm` are same cost for most/all ops other than `mov`).

eugeneia commented:
I have adopted some of the changes suggested by @corsix, which show quite a substantial improvement in performance. I have omitted the loop rotations, since they didn’t seem to make a measurable difference when testing and make the code harder to read imho.

I also didn’t do the BMI2 dependent changes since the oldest lab server doesn’t support them (grindelwald has an ivy bridge CPU), also I would have to add the BMI2 instructions to DynASM first. Might be worth to do a CPU feature detection here?

Before:
```
PMU analysis (numentries=400000, keysize=64)
build: 6.7974 seconds
lookup64: 576.12 cycles/lookup 215.73 instructions/lookup
build(direct_pointing): 7.3970 seconds
lookup64(direct_pointing): 441.36 cycles/lookup 150.59 instructions/lookup
```
After:
```
PMU analysis (numentries=400000, keysize=64)
build: 6.9591 seconds
lookup64: 507.08 cycles/lookup 186.47 instructions/lookup
build(direct_pointing): 7.5430 seconds
lookup64(direct_pointing): 404.69 cycles/lookup 132.36 instructions/lookup
```

corsix commented:
FWIW, the latest upstream DynASM should have BMI2 in it (c.f. https://github.com/LuaJIT/LuaJIT/blob/26f1023819efb843e10014232cd88bb1d52ea4f5/dynasm/dasm_x86.lua#L1690, present there since February 2018). Not having Haswell everywhere is a valid reason not to use it though (I would however be interested in knowing what performance improvement - if any - it could give here, if that experiment is something you're willing to try).

corsix commented on src/lib/poptrie_lookup.dasl:
```diff
@@ -0,0 +1,146 @@
+   -- rax = popcnt(band(vec, rax)) - 1
+   | and rax, vec
+   | popcnt rax, rax
+   -- return leaves[base + bc - 1]
```
The `mov` and the `add` here could be fused, as in:
```
   | add eax, dword [node+16] -- nodes[index].base0
   | movzx eax, word [leaves+eax*2-2] -- leaves[index]
```
(No improvement in unfused-domain uops, but does save one fused-domain uop per function invocation, if that is even measurable against noise)

corsix commented:
Very nice; I like seeing concrete performance numbers.

eugeneia commented:
@petebristow I am having trouble patching in the BMI2 instructions from https://github.com/LuaJIT/LuaJIT/commit/fe651bf6e2b4d02b624be3c289378c08bab2fa9b

I get errors like

```
lib/poptrie_lookup.dasl:98: error: bad char `w' in pattern `F20F38wF7rM' for `shrx':
      | shrx v, key, offset
lib/poptrie_lookup.dasl:109: error: bad char `w' in pattern `0F38wF5rM' for `bzhi':
   | bzhi rax, vec, rcx
lib/poptrie_lookup.dasl:130: error: bad char `w' in pattern `0F38wF5rM' for `bzhi':
   | bzhi rax, vec, rcx
```
 and can’t seem to find where the `w` modifier is actually implemented?

corsix commented:
Aha, the `w` got lost as [the PR](https://github.com/LuaJIT/LuaJIT/pull/399/files) became [the upstream commit](https://github.com/LuaJIT/LuaJIT/commit/fe651bf6e2b4d02b624be3c289378c08bab2fa9b). That is unfortunate.

eugeneia commented:
@corsix On another note, I am currently updating this code to work with keys up to 128 bits (IPv6) and am wondering if you know any neat tricks for extracting 6 bits from a `uint8_t[16]` at an arbitrary offset in x86 asm? (I suppose what is currently an `uint64_t` argument will have to become a pointer.)

corsix commented:
shrd from a pair of uint64_t registers comes to mind. Not a cheap
instruction (e.g. four uops on Broadwell), but should do what you need.

On Fri, 8 Feb 2019 at 08:40, Max Rottenkolber <notifications@github.com>
wrote:

> @corsix <https://github.com/corsix> On another note, I am currently
> updating this code to work with keys up to 128 bits (IPv6) and am wondering
> if you know any neat tricks for extracting 6 bits from a uint8_t[16] at
> an arbitrary offset in x86 asm? (I suppose what is currently an uint64_t
> argument will have to become a pointer.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/snabbco/snabb/pull/1397#issuecomment-461729686>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGhlWhgvHiBE7hW2JajDkcsGFQxDvJks5vLTf7gaJpZM4ZTsDc>
> .
>

eugeneia commented:
Without BMI2 (this branch):
```
PMU analysis (numentries=400000, keysize=64)
build: 6.7322 seconds
lookup: 55733.43 cycles/lookup 87694.39 instructions/lookup
lookup64: 504.83 cycles/lookup 186.46 instructions/lookup
build(direct_pointing): 7.5217 seconds
lookup(direct_pointing): 11152.73 cycles/lookup 25908.96 instructions/lookup
lookup64(direct_pointing): 416.15 cycles/lookup 132.36 instructions/lookup
```
With BMI2 (https://github.com/eugeneia/snabb/commits/poptrie-bmi2):
```
PMU analysis (numentries=400000, keysize=64)
build: 7.3056 seconds
lookup: 39330.65 cycles/lookup 72852.83 instructions/lookup
lookup64: 431.63 cycles/lookup 151.47 instructions/lookup
build(direct_pointing): 7.8106 seconds
lookup(direct_pointing): 12443.39 cycles/lookup 26317.73 instructions/lookup
lookup64(direct_pointing): 376.19 cycles/lookup 112.10 instructions/lookup
```

eugeneia commented:
@corsix Thanks for the suggestion! I ended up with something like this for extracting from a 16 byte array, probably not optimal... wdyt?

```
mov rxc, offset
mov rax, [key]
mov rbx, [key+8]
cmp rcx, 63
jng >1
sub rcx, 64
mov rax, rbx
1:
shrd rax, rbx, cl
and rax, 0x3F
```

corsix commented:
```
mov ecx, offset   -- Prefer 32-bit operations where possible
mov rax, [key]
mov rbx, [key+8]
test cl, 64       -- Prefer test to cmp (archaic habbit)
cmovnz rax, rbx   -- This is your mov within the branch
                  -- Do not need to sub rcx, 64
                  -- as shrd takes cl mod 64, and subtracting 64 mod 64 is a no-op
shrd rax, rbx, cl
and eax, 0x3F     -- Prefer 32-bit operations where possible
```

eugeneia commented:
@corsix Neat! Thanks!

I have added support for 128 (and 32) bit keys (the interface now expects pointers to byte arrays for keys), a well as conditional use of BMI2 instructions (since they have a substantial impact.)

I have adjusted the micro benchmark so that it does not measure a cold cache. It is not exactly representative of a real workload either, but yields more accurate feedback for testing changes. I have also noticed that for 400,000 entries the benchmark will now OOM, but could confirm that this will be fixed by switching to RaptorJIT.

I think the results tell us that its pretty fast with warm caches, and cost increases roughly linearly with key size (prefix length).

```
PMU analysis (numentries=400000, keysize=32)
build: 6.0666 seconds
lookup: 5604.90 cycles/lookup 14653.72 instructions/lookup
lookup32: 81.79 cycles/lookup 118.72 instructions/lookup
lookup64: 81.77 cycles/lookup 118.72 instructions/lookup
lookup128: 99.13 cycles/lookup 138.95 instructions/lookup
build(direct_pointing): 6.0043 seconds
lookup(direct_pointing): 2605.19 cycles/lookup 4805.54 instructions/lookup
lookup32(direct_pointing): 46.62 cycles/lookup 79.66 instructions/lookup
lookup64(direct_pointing): 47.05 cycles/lookup 79.65 instructions/lookup
lookup128(direct_pointing): 51.88 cycles/lookup 84.86 instructions/lookup
```
```
PMU analysis (numentries=400000, keysize=64)
build: 17.5607 seconds
lookup: 48630.60 cycles/lookup 58459.67 instructions/lookup
lookup64: 131.99 cycles/lookup 153.46 instructions/lookup
lookup128: 144.89 cycles/lookup 185.23 instructions/lookup
build(direct_pointing): 17.5885 seconds
lookup(direct_pointing): 4672.99 cycles/lookup 12343.01 instructions/lookup
lookup64(direct_pointing): 92.95 cycles/lookup 114.37 instructions/lookup
lookup128(direct_pointing): 96.92 cycles/lookup 131.13 instructions/lookup
```

eugeneia commented:
I’ll be my own upstream for this and I think this is ready, merged into max-next.
